### PR TITLE
Allow Java Harness to understand Fns the Java Runner produces.

### PR DESCRIPTION
**Pull request by robertwb**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java [hunk 1]
```diff
@@ -68,6 +68,7 @@
 
   // TODO: What should the initial set of URNs be?
   private static final String DATA_INPUT_URN = "urn:org.apache.beam:source:runner:0.1";
+  public static final String READ_URN = "urn:beam:transform:read:v1";
```
Comment (kennknowles) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java]: Can these URNs be imported from e.g. PTransformTranslation? We also do need to get them in a central place, I know, but anyhow in Java we have this available.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java [hunk 2]
```diff
@@ -137,6 +137,7 @@ public void processInstructionRequests(Executor executor)
       executor.execute(
           () -> sendInstructionResponse(delegateOnInstructionRequestType(currentRequest)));
     }
+    Thread.sleep(50);
```
Comment (kennknowles) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java]: Comment as to why?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java [hunk 3]
```diff
@@ -109,7 +114,9 @@
 
     @Override
     public Map<String, PTransformRunnerFactory> getPTransformRunnerFactories() {
-      return ImmutableMap.of(ParDoTranslation.CUSTOM_JAVA_DO_FN_URN, new Factory());
+      return ImmutableMap.of(
+          PTransformTranslation.PAR_DO_TRANSFORM_URN, new NewFactory(),
```
Comment (kennknowles) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java]: Did you name this just for the delightful `new NewFactory()`?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BoundedSourceRunner.java [hunk 4]
```diff
@@ -124,10 +128,17 @@ public void start() throws Exception {
...
-      InputT boundedSource =
-          (InputT) SerializableUtils.deserializeFromByteArray(bytes, definition.toString());
+      InputT boundedSource;
+      if (definition.getUrn().equals(ProcessBundleHandler.JAVA_SOURCE_URN)) {
+        byte[] bytes = definition.getPayload().toByteArray();
+        @SuppressWarnings("unchecked")
+        InputT boundedSource0 =
+            (InputT) SerializableUtils.deserializeFromByteArray(bytes, definition.toString());
+        boundedSource = boundedSource0;
+      } else {
```
Comment (kennknowles) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BoundedSourceRunner.java]: It kind of looks like this branch is expected to have READ_URN. Is that so? Does that not work? I'm just curious, and would love to have the bare `else` case be an error, with all understood cases being guarded by a URN.

File: sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BoundedSourceRunnerTest.java [hunk 5]
```diff
@@ -146,6 +146,7 @@ public void testCreatingAndProcessingSourceFromFactory() throws Exception {
         Suppliers.ofInstance("57L")::get,
         ImmutableMap.of(),
         ImmutableMap.of(),
+        ImmutableMap.of(),
```
Comment (kennknowles) [inline sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BoundedSourceRunnerTest.java]: Nit: `Collections.emptyMap()` is even more clear, despite the surrounding code.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BoundedSourceRunner.java (hunk 4 above)
Comment (robertwb) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BoundedSourceRunner.java]: That is correct (given the factory registration above), but made more explicit.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java (hunk 3 above)
Comment (robertwb) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java]: :). I just needed a new name for the new factory. The existing one will go away eventually.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java (hunk 2 above)
Comment (robertwb) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java]: Leftover debugging. (There used to be some logging and flushing here as well.) Removed.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java (hunk 1 above)
Comment (robertwb) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java]: There is a central place :-P https://github.com/apache/beam/blob/master/sdks/python/apache_beam/utils/urns.py 

Used (existing!) READ_URN from PTransformTranslation. The others are legacy/bundle specific, leaving for now.

File: sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BoundedSourceRunnerTest.java (hunk 5 above)
Comment (robertwb) [reply sdks/java/harness/src/test/java/org/apache/beam/fn/harness/BoundedSourceRunnerTest.java]: Changed.

**Comment by robertwb** [comment]
Yes, this grew out of trying to use the PULR, but I knew we needed for the JULR and will want to migrate the harness to these as well.

**Comment by robertwb** [comment]
Thanks for the review, PTAL.

**Comment by robertwb** [comment]
@kennknowles Any more thoughts on this?

**Comment by kennknowles** [comment]
LGTM; you'll need to rebase to resolve conflicts before you can clicky merge. Definitely useful to move forward here. Sorry for the delay.

**Pull request merged**